File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class WordDictionary {
    private TrieNode root;

    public WordDictionary() {
        this.root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode cur = root;

        for (char c : word.toCharArray()) {
            if (cur.children[c - 'a'] == null) {
                cur.children[c - 'a'] = new TrieNode();
            }

            cur = cur.children[c - 'a'];
        }
        
        cur.isEndOfWord = true;
    }
    
    public boolean search(String word) {
        return dfs(root, word, 0);
    }

    private boolean dfs(TrieNode node, String word, int index) {
        if (index == word.length()) return node.isEndOfWord;

        char c = word.charAt(index);

        if (c == '.') {
            for (TrieNode child : node.children) {
                if (child != null) {
                    if (dfs(child, word, index + 1)) {
                        return true;
                    }
                }
            }

            return false;
        } else {
            if (node.children[c - 'a'] == null) {
                return false;
            }

            return dfs(node.children[c - 'a'], word, index + 1);
        }
    }

    private class TrieNode {
        TrieNode[] children;
        boolean isEndOfWord;

        public TrieNode() {
            this.children = new TrieNode[26];
            this.isEndOfWord = false;
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.kt
class WordDictionary() {
    private val root = TrieNode()

    fun addWord(word: String) {
        var cur = root

        for (c in word) {
            val index = c - 'a'

            if (cur.children[index] == null) {
                cur.children[index] = TrieNode()
            }

            cur = cur.children[index]!!
        }

        cur.isEndOfWord = true
    }

    fun search(word: String): Boolean {
        return dfs(root, word, 0)
    }

    private fun dfs(node: TrieNode, word: String, index: Int): Boolean {
        if (index == word.length) return node.isEndOfWord

        val c = word[index]

        return if (c == '.') {
            node.children.any { child ->
                child != null && dfs(child, word, index + 1)
            }
        } else {
            val childIndex = c - 'a'
            node.children[childIndex]?.let { dfs(it, word, index + 1) } ?: false
        }
    }

    private class TrieNode {
        val children = arrayOfNulls<TrieNode>(26)
        var isEndOfWord = false
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */